datafusion/optimizer/src/replace_distinct_aggregate.rs

```diff
@@ -88,60 +94,72 @@ impl OptimizerRule for ReplaceDistinctWithAggregate {
                 input,
                 schema,
             })) => {
+                let expr_cnt = on_expr.len();
+
                 // Construct the aggregation expression to be used to fetch the selected expressions.
-                let aggr_expr = select_expr
-                    .iter()
-                    .map(|e| {
-                        Expr::AggregateFunction(AggregateFunction::new(
-                            AggregateFunctionFunc::FirstValue,
-                            vec![e.clone()],
-                            false,
-                            None,
-                            sort_expr.clone(),
-                            None,
-                        ))
-                    })
-                    .collect::<Vec<Expr>>();
+                let aggr_expr = vec![Expr::AggregateFunction(AggregateFunction::new(
+                    AggregateFunctionFunc::FirstValue,
+                    select_expr,
+                    false,
+                    None,
+                    sort_expr.clone(),
+                    None,
+                ))];
+
+                let aggr_expr = normalize_cols(aggr_expr, input.as_ref())?;
+                let group_expr = normalize_cols(on_expr, input.as_ref())?;
 
                 // Build the aggregation plan
-                let plan = LogicalPlanBuilder::from(input.as_ref().clone())
-                    .aggregate(on_expr.clone(), aggr_expr.to_vec())?
-                    .build()?;
+                let plan = LogicalPlan::Aggregate(Aggregate::try_new(
+                    input, group_expr, aggr_expr,
+                )?);
+                let lpb = LogicalPlanBuilder::from(plan);
```

Comment:
Shall we make LogicalPlanBuilder accept AsRef<LogicalPlan> or Into<Arc<LogicalPlan>>?

Reply:
I don't understand this question. This code looks good to me (like it won't copy `plan`)

Reply:
The context is that, I extract this logic totally from `LogicalPlanBuilder::aggregate`. It might introduce a problem that if `LogicalPlanBuilder::aggregate` changes in the future, the code here stays still and frustrates people to find and rewrite.

The reason why I extract the logic, is that the `LogicalPlanBuilder` accepts a `LogicalPlan`. So that when you'd like to build an aggregate plan with `LogicalPlanBuilder`, it will go through:

```rust
// the builder accepts LogicalPlan only, so a clone needed
let mut builder = LogicalPlanBuilder::from(Arc<LogicalPlan>.as_ref().clone());
// then you get:
builder == LogicalPlanBuilder { LogicalPlan };

// making an aggregate plan
builder.aggregate(..);
// this will call Arc::new() on the LogicalPlan contained.
// another clone here
LogicalPlanBuilder { AggrPlan { Arc<LogicalPlan>, ... } }
```

Since many of our `LogicalPlan`s could contain an `Arc<LogicalPlan>`, and for such occasion, I wonder if `Into<Arc<LogicalPlan>>` could save clones, while still allows us to use `LogicalPlanBuilder`, instead of extracting logic.

```rust
struct LogicalPlanBuilder<PB> {
    plan: PB,
}
impl<PB: Into<Arc<LogicalPlan>>> From<PB> for LogicalPlanBuilder<PB> {
    // ...
}
```
Because `Arc<T>` has `From<T>` implemented, so when we need `Arc::new`, just call `into` on `self.plan`, this should save times of clones.

Reply:
Ah, I think you are asking about getting `LogicalPlan` from `Arc<LogicalPlan>` 

One way to do it today is using `unwrap_arc` (source below) which will only copy when necessary

```rust
// the builder accepts LogicalPlan only, so a clone needed
let mut builder = LogicalPlanBuilder::from(unwrap_arc(plan));

// making an aggregate plan
builder.aggregate(..);
// this will call Arc::new() on the LogicalPlan contained.
```

https://github.com/apache/datafusion/blob/1eff714ef8356dc305047386ba250b62bed6a795/datafusion/expr/src/logical_plan/tree_node.rs#L368



Separately, adding an `LogicalPlanBuilder::from(input_plan_arc)` would be a great idea. I'll file a ticket about doing so

Reply:
filed https://github.com/apache/datafusion/issues/10465

Reply:
```suggestion
                // TODO use LogicalPlanBuilder directly rather than recreating the Aggregate
                // when https://github.com/apache/datafusion/issues/10485 is available
                let lpb = LogicalPlanBuilder::from(plan);
```

I filed https://github.com/apache/datafusion/issues/10485 to track this. I think we could merge this PR as is (I added a comment) and improve the code as a follow on PR